protoc-gen-gogo/grpcnamed/grpc.go

```diff
@@ -0,0 +1,480 @@
+func (g *grpc) generateService(file *generator.FileDescriptor, service *pb.ServiceDescriptorProto, index int) {
+	path := fmt.Sprintf("6,%d", index) // 6 means service.
+
+	origServName := service.GetName()
+	fullServName := origServName
+	if pkg := file.GetPackage(); pkg != "" {
+		fullServName = pkg + "." + fullServName
+	}
+	servName := generator.CamelCase(origServName)
+
+	g.P()
+	g.P("// Client API for ", servName, " service")
+	g.P()
+
+	// Client interface.
+	g.P("type ", servName, "Client interface {")
+	for i, method := range service.Method {
+		g.gen.PrintComments(fmt.Sprintf("%s,2,%d", path, i)) // 2 means method in a service.
+		g.P(g.generateClientSignature(servName, method))
+	}
+	g.P("}")
+	g.P()
+
+	// Client structure.
+	g.P("type ", unexport(servName), "Client struct {")
+	g.P("cc *", grpcPkg, ".ClientConn")
+	g.P("name string")
+	g.P("}")
+	g.P()
+
+	// NewClient factory.
+	g.P("func NewNamed", servName, "Client (name string, cc *", grpcPkg, ".ClientConn) ", servName, "Client {")
+	g.P("return &", unexport(servName), "Client{cc:cc, name:name}")
+	g.P("}")
+	g.P()
+	g.P("func New", servName, "Client (cc *", grpcPkg, ".ClientConn) ", servName, "Client {")
+	g.P("return NewNamed", servName, "Client(\"", fullServName, "\", cc)")
+	g.P("}")
+	g.P()
+
+	var methodIndex, streamIndex int
+	methodsDescVar := "_" + servName + "_methodsDesc"
+	streamsDescVar := "_" + servName + "_streamsDesc"
+	// Client method implementations.
+	for _, method := range service.Method {
+		var descExpr string
+		if !method.GetServerStreaming() && !method.GetClientStreaming() {
+			// Unary RPC method
+			descExpr = fmt.Sprintf("&%s[%d]", methodsDescVar, methodIndex)
+			methodIndex++
+		} else {
+			// Streaming RPC method
+			descExpr = fmt.Sprintf("&%s[%d]", streamsDescVar, streamIndex)
+			streamIndex++
+		}
+		g.generateClientMethod(servName, method, descExpr)
+	}
+
+	g.P("// Server API for ", servName, " service")
+	g.P()
+
+	// Server interface.
+	serverType := servName + "Server"
+	g.P("type ", serverType, " interface {")
+	for i, method := range service.Method {
+		g.gen.PrintComments(fmt.Sprintf("%s,2,%d", path, i)) // 2 means method in a service.
+		g.P(g.generateServerSignature(servName, method))
+	}
+	g.P("}")
+	g.P()
+
+	g.P("func _named", servName, "ServiceDesc(name string) *", grpcPkg, ".ServiceDesc {")
+	g.P("return &", grpcPkg, ".ServiceDesc{")
+	g.P("ServiceName: name,")
+	g.P("HandlerType: (*", serverType, ")(nil),")
+	g.P("Methods: ", methodsDescVar, ",")
+	g.P("Streams: ", streamsDescVar, ",")
+	g.P("Metadata: \"", file.GetName(), "\",")
+	g.P("}")
+	g.P("}")
+	g.P()
+	// Server registration.
+	g.P("func RegisterNamed", servName, "Server(s *", grpcPkg, ".Server, srv ", serverType, ", name string) {")
+	g.P("s.RegisterService(_named", servName, "ServiceDesc(name), srv)")
+	g.P("}")
+	g.P()
+	g.P("func Register", servName, "Server(s *", grpcPkg, ".Server, srv ", serverType, ") {")
```

Comment:
I thought that this plugin would only generate the RegisterNamed functions and that you won't need to generate the other Register functions, since they will be generated by the original grpc plugin.
This plugin should only need to do the extra code generation, which means it should require much less code.